Zend/zend_compile.c

```diff
@@ -5158,6 +5158,250 @@ void zend_compile_switch(zend_ast *ast) /* {{{ */
 }
 /* }}} */
 
+static uint32_t count_match_conds(zend_ast_list *arms)
+{
+	uint32_t num_conds = 0;
+
+	for (uint32_t i = 0; i < arms->children; i++) {
+		zend_ast *arm_ast = arms->child[i];
+		if (arm_ast->child[0] == NULL) {
+			continue;
+		}
+
+		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+		num_conds += conds->children;
+	}
+
+	return num_conds;
+}
+
+static zend_uchar determine_match_jumptable_type(zend_ast_list *arms) {
+	zend_uchar common_type = IS_UNDEF;
+	for (uint32_t i = 0; i < arms->children; i++) {
+		zend_ast *arm_ast = arms->child[i];
+		if (!arm_ast->child[0]) {
+			/* Skip default clause */
+			continue;
+		}
+
+		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+		for (uint32_t j = 0; j < conds->children; j++) {
+			zend_ast **cond_ast = &conds->child[j];
+
+			zend_eval_const_expr(cond_ast);
+			if ((*cond_ast)->kind != ZEND_AST_ZVAL) {
+				/* Non-constant case */
+				return IS_UNDEF;
+			}
+
+			zval *cond_zv = zend_ast_get_zval(*cond_ast);
+			if (Z_TYPE_P(cond_zv) != IS_LONG && Z_TYPE_P(cond_zv) != IS_STRING) {
+				/* We only optimize switched on integers and strings */
+				return IS_UNDEF;
+			}
+
+			if (common_type == IS_UNDEF) {
+				common_type = Z_TYPE_P(cond_zv);
+			} else if (common_type != Z_TYPE_P(cond_zv)) {
+				/* Non-uniform case types */
+				return IS_UNDEF;
+			}
+
+			if (Z_TYPE_P(cond_zv) == IS_STRING
```

Comment:
I think this block can just be removed - In PHP, `'00' == '0'` is true, which is a concern for `switch`, but that isn't a concern for `match`, which would check if values are identical(===) instead of equal(==)

nit: Comments (`/* Non-uniform case types */`, `/* We only optimize switched on integers and strings */`, `/* Non-constant case */` should refer to arms instead of cases, match instead of switch.